### PR TITLE
Solution

### PR DESCRIPTION
**Pull request by PastushenkoVik**

### CODE REVIEW

File: src/averageAges.js [hunk 1]
```diff
@@ -15,11 +15,14 @@
  * @return {number}
  */
 function calculateMenAverageAge(people, century) {
-  // write code here
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with logical operators (&&, ||) or ternary operator (?:)
-  // without nesting
+  const men = people
+    .filter(item =>
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: do not break line just after `=>`
it is not convenient, hard to read and I think may cause bugs during interpretation, like 
```
if (true) console.log('a')
console.log('b');
```
but I am not shure

File: src/averageAges.js [hunk 2]
```diff
@@ -15,11 +15,14 @@
...
+      item.sex === 'm'
+      && (century === undefined
+        || (item.died < century * 100 && item.died >= (century - 1) * 100)))
+    .map(item => item.died - item.born);
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: Your code looks intuitive, but you iterate array 3 times, can you move all code to reduce?

File: src/averageAges.js [hunk 3]
```diff
@@ -34,7 +37,14 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people
+    .filter(item =>
+      item.sex === 'f'
+      && (withChildren === undefined
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: explicite is good, you do not must change it, but more convenient just `!withChildren` if you do not need differences between `false`, `null`, `undefined`, etc.

File: src/averageAges.js [hunk 4]
```diff
@@ -34,7 +37,14 @@ function calculateMenAverageAge(people, century) {
...
+        || people.find(child => child.mother === item.name) !== undefined))
+    .map(item => item.died - item.born);
+
+  return women.reduce((sum, item) => sum + item, 0) / women.length;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: the same comments as in first function

File: src/averageAges.js [hunk 5]
```diff
@@ -34,7 +37,14 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const women = people
+    .filter(item =>
+      item.sex === 'f'
+      && (withChildren === undefined
+        || people.find(child => child.mother === item.name) !== undefined))
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: can you rewrite without nested loops

File: src/averageAges.js [hunk 6]
```diff
@@ -52,11 +62,17 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const children = people
+    .filter(child => (onlyWithSon === undefined || child.sex === 'm')
+      && people.find(parent => parent.name === child.mother) !== undefined)
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: nested loop

Comment (YuriiHoliuk) [reply src/averageAges.js]: Also, there are useful `some` method

File: src/averageAges.js [hunk 7]
```diff
@@ -52,11 +62,17 @@ function calculateWomenAverageAge(people, withChildren) {
...
+    .map(child =>
+      child.born - people.find(mother => mother.name === child.mother).born);
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: nested loop again, to find the same person, try to optimize performance

File: src/averageAges.js [hunk 8]
```diff
@@ -52,11 +62,17 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+  return children.reduce((sum, item) => sum + item, 0) / children.length;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: Maybe, can you reuse this part or bigger

**Review by YuriiHoliuk — CHANGES_REQUESTED** [review]

File: src/averageAges.js (hunk 1 above)
Comment (PastushenkoVik) [reply src/averageAges.js]: In this case we have only conditions, then I want indent them at the same column
I tried another indent. you will see

File: src/averageAges.js [hunk 9]
```diff
@@ -15,11 +15,19 @@
...
-  // avoid using loop and forEach
-  // replace `if ()` statement with logical operators (&&, ||) or ternary operator (?:)
-  // without nesting
+  let counter = 0;
+
+  return people
+    .reduce((ageSum, man) => {
+      if (man.sex === 'm' && (century === undefined
+        || (man.died < century * 100 && man.died >= (century - 1) * 100))) {
+        counter++;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: mutate variables outside callback(map, reduce) is bad practice

File: src/averageAges.js [hunk 10]
```diff
@@ -15,11 +15,19 @@
...
-  // learn how to use array methods like .filter .map .some .every .find .reduce
-  // avoid using loop and forEach
-  // replace `if ()` statement with logical operators (&&, ||) or ternary operator (?:)
-  // without nesting
+  let counter = 0;
+
+  return people
+    .reduce((ageSum, man) => {
+      if (man.sex === 'm' && (century === undefined
+        || (man.died < century * 100 && man.died >= (century - 1) * 100))) {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: hard to read, but ok

File: src/averageAges.js [hunk 11]
```diff
@@ -34,7 +42,21 @@ function calculateMenAverageAge(people, century) {
...
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const mothers = people.map(item => item.mother);
+
+  let counter = 0;
+
+  return people
+    .reduce((ageSum, woman) => {
+      if (woman.sex === 'f'
+        && (withChildren === undefined || mothers.includes(woman.name))) {
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: includes also have loop inside

File: src/averageAges.js [hunk 12]
```diff
@@ -52,11 +74,25 @@ function calculateWomenAverageAge(people, withChildren) {
  * @return {number}
  */
 function calculateAverageAgeDiff(people, onlyWithSon) {
-  // write code here
+  const mothers = new Map(people.map(item => [item.name, item.born]));
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: This is intuitive to use Map, but plain object will be enough, will work faster and has shorter syntax

**Review by YuriiHoliuk — CHANGES_REQUESTED** [review]
Стосовно цього може і ок
Колбек мав би бути чистою функцією, але без цього буде більше коду, як хочеш

File: src/averageAges.js [hunk 13]
```diff
@@ -15,11 +15,15 @@
...
-  // without nesting
+  return people.reduce((ageSum, man) => {
+    if (man.sex === 'm' && (century === undefined
+      || (man.died < century * 100 && man.died >= (century - 1) * 100))) {
+      return ageSum + man.died - man.born;
+    }
+
+    return ageSum;
+  }, 0) / people.filter(man => (man.sex === 'm' && (century === undefined
+    || (man.died < century * 100 && man.died >= (century - 1) * 100)))).length;
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: you duplicate a lot of code, also use unneeded filter
Proposed solution for reduce only:
```javascript
people.reduce(({ sum, count }, person) => {
  if (...) {
    return { sum: sum + ..., count: count + 1 }
  }
  return { sum, count };
}, { sum: 0, count: 0 })
```

Your previous solution with external variable was better then current
Or you can just use for loop

File: src/averageAges.js [hunk 14]
```diff
@@ -34,7 +38,22 @@ function calculateMenAverageAge(people, century) {
...
+
+  return people.reduce((ageSum, woman) => {
+    if (woman.sex === 'f'
+      && (withChildren === undefined
+        || mothers[woman.name] !== undefined)) {
+      return ageSum + woman.died - woman.born;
+    }
+
+    return ageSum;
+  }, 0) / people.filter(woman => woman.sex === 'f'
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: the same

File: src/averageAges.js [hunk 15]
```diff
@@ -52,11 +71,25 @@ function calculateWomenAverageAge(people, withChildren) {
...
+
+  return people.reduce((age, child) => {
+    if ((onlyWithSon === undefined || child.sex === 'm')
+      && mothers[child.mother] !== undefined) {
+      return age + child.born - mothers[child.mother];
+    }
+
+    return age;
+  }, 0) / people
+    .filter(child => (onlyWithSon === undefined || child.sex === 'm')
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: the same

**Review by YuriiHoliuk — CHANGES_REQUESTED** [review]

File: src/averageAges.js [hunk 16]
```diff
@@ -34,7 +36,17 @@ function calculateMenAverageAge(people, century) {
  * @return {number}
  */
 function calculateWomenAverageAge(people, withChildren) {
-  // write code here
+  const mothers = people.reduce((obj, item) => {
+    obj[item.mother] = '';
```
Comment (YuriiHoliuk) [inline src/averageAges.js]: will be simpler with `obj[item.mother] = true;`

**Review by YuriiHoliuk — APPROVED** [review]

**Pull request closed without merging**